### PR TITLE
Allow preferences to be an array of serializable values.

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
Allow preferences to be an array of serializable values.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/16594

#### Docs PR
In progress

### CODE REVIEW

**Review by skabashnyuk — APPROVED** [review]

**Comment by mshaposhnik** [comment]
[ci-test]

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/SerializableConverter.java [hunk 1]
```diff
@@ -47,7 +49,8 @@ public String convertToDatabaseColumn(Serializable value) {
   @Override
   public Serializable convertToEntityAttribute(String dbData) {
     try {
-      return objectMapper.readValue(dbData, Serializable.class);
+      Serializable[] arr = objectMapper.readValue(dbData, Serializable[].class);
+      return (arr.length == 1) ? arr[0] : arr;
```
Comment (metlos) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/SerializableConverter.java]: What about legit 1 element arrays?

**Review by metlos — COMMENTED** [review]
I think we should be able to distinguish between a value and an 1 element array...

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/SerializableConverter.java (hunk 1 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/SerializableConverter.java]: Good question..
Any ideas better than:
```
try {
  objectMapper.readValue(dbData, Serializable[].class);
} catch (Exc e) {  //not an array
  objectMapper.readValue(dbData, Serializable.class);
}
```
?

Comment (metlos) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/SerializableConverter.java]: What about reading it as json nodes and see if it is an array node or a value node?

Comment (mshaposhnik) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/SerializableConverter.java]: little overhead, but probably will work..

**Comment by mshaposhnik** [comment]
crw-ci-test

**Comment by mshaposhnik** [comment]
crw-ci-test

**Comment by bfitzpat** [comment]
Hi folks! I have noticed a weird issue with processing of JSON arrays as preference values passed in a devfile. Spaces after commas seems to hang things up for me. 
i.e. this hangs on processing
``` didact.append.registry: '[{"name":"AppendMe", "category":"AppendedCat", "sourceUri":"https://raw.githubusercontent.com/redhat-developer/vscode-didact/master/examples/registry.example.didact.md"}]'```
but this does not
```didact.append.registry: '[{"name":"AppendMe","category":"AppendedCat","sourceUri":"https://raw.githubusercontent.com/redhat-developer/vscode-didact/master/examples/registry.example.didact.md"}]'```
Is this a known issue? Or am I just doing something wrong?

**Comment by sparkoo** [comment]
@bfitzpat Hi, I don't think it is known issue. Can you please create a bug issue and fill in all questions from the issue template? https://github.com/eclipse/che/issues/new/choose
thanks

**Pull request merged**